packages/opal-common/opal_common/authentication/authenticator.py

```diff
@@ -0,0 +1,62 @@
+from abc import abstractmethod
+from typing import Optional
+
+from fastapi import Header
+from opal_common.authentication.deps import JWTAuthenticator
+from opal_common.authentication.types import JWTClaims
+from opal_common.authentication.verifier import JWTVerifier, Unauthorized
+from opal_common.config import opal_common_config
+from opal_common.logger import logger
+
+from .oauth2 import CachedOAuth2Authenticator, OAuth2ClientCredentialsAuthenticator
+
+
+class Authenticator:
+    @property
+    def enabled(self):
+        return self._delegate().enabled
+
+    async def authenticate(self, headers):
+        if hasattr(self._delegate(), "authenticate") and callable(
+            getattr(self._delegate(), "authenticate")
+        ):
+            await self._delegate().authenticate(headers)
+
+    @abstractmethod
+    def _delegate(self) -> dict:
+        pass
+
+
+class _ClientAuthenticator(Authenticator):
+    def __init__(self):
+        if opal_common_config.AUTH_TYPE == "oauth2":
```

Comment:
Instead of having this class "proxy" the actual implementation according to the configured `AUTH_TYPE` - you probably want to use the `factory` design pattern.
Have both `CachedOAuth2Authenticator` & `_JWTAuthenticator` inherit from `Authenticator` and implement its declared methods.

Then have another class / function (e.g `AuthenticatorFactory.create`) with return type `Authenticator` instantiate the right subclass according to configured auth type (or other relevant arguments).

Basically what you're doing here is having a class "deciding what it is" within its constructor, which again doesn't make use of Python's native polymorphism / OOP support.

Reply:
Modified code to use `AuthenticatorFactory`.